### ISSUE TITLE
Metal backend does not properly release image memory

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#61781]
Tested on engine revision flutter/engine@d3b81f19fc98076283892c1152d30e2567b6406e

run the attached program below at framework revision 45e097da16

Press the `LOAD IMAGE` button, note memory usage increase. Press the `CLEAR IMAGE` button. Press the `CHECK ALLOC PROFILE (Try GC)`. Note that memory does not decrease.

Build the engine locally, but use the Open GL backend (set `shell_enable_metal` to false and `skia_use_metal` to false). Follow the same steps - memory gets cleared when GC runs.

This is a problem because subsequent image allocations will bump the memory up temporarily - with enough or sufficiently large allocations, this can easily lead to an OOM that is avoidable.

/cc @jvanverth @chinmaygarde 

This program:

```dart
import 'dart:async';
import 'dart:developer' as developer;
import 'dart:isolate';
import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:vm_service/vm_service.dart' hide Isolate;
import 'package:vm_service/vm_service_io.dart';

const int width = 8000;
const int height = 8000;
const int bytes = width * height * 4;
Uint8List pixels;
VmService vmService;
String isolateId;

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  imageCache.maximumSize = 0;
  pixels = await createPixels(bytes);

  final developer.ServiceProtocolInfo info = await developer.Service.getInfo();
  isolateId = developer.Service.getIsolateID(Isolate.current);
  if (info.serverUri == null) {
    throw Exception('This app must be run in debug or profile mode');
  }

  vmService = await vmServiceConnectUri(
      'ws://localhost:${info.serverUri.port}${info.serverUri.path}ws');

  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

Future<Uint8List> createPixels(int bytes) async {
  final pixels = Uint8List(bytes);
  for (int index = 0; index < bytes; index += 4) {
    pixels[index++] = index % 128;
    pixels[index++] = index % 127;
    pixels[index++] = index % 126;
    pixels[index++] = 0xFF;
  }
  return pixels;
}

Future<ui.Image> createImage() async {
  final completer = Completer<ui.Image>();

  ui.decodeImageFromPixels(
    pixels,
    width,
    height,
    ui.PixelFormat.rgba8888,
    (ui.Image image) => completer.complete(image),
  );
  return completer.future;
}

class _MyHomePageState extends State<MyHomePage> {
  ui.Image image;
  bool loadingImage = false;
  String imageButtonText = 'LOAD IMAGE';
  String memoryUsageLabel = 'No memory usage recorded.';

  void toggleImage() {
    assert(!loadingImage);
    setState(() {
      if (image == null) {
        loadingImage = true;
        imageButtonText = 'Loading image...';
        createImage().then((ui.Image image) {
          setState(() {
            imageButtonText = 'CLEAR IMAGE';
            loadingImage = false;
            this.image = image;
          });
        });
      } else {
        setState(() {
          image = null;
          imageButtonText = 'LOAD IMAGE';
        });
      }
    });
  }

  VoidCallback sendLowMemory(bool gc) {
    // GC is not for sure, but usually works.
    // Also note that even _without_ trying to trigger a GC, this
    // creates enough new garbage to sometimes cause a GC anyway.
    return () {
      vmService.getAllocationProfile(isolateId, gc: gc).then((dynamic value) {
        AllocationProfile profile = value as AllocationProfile;
        setState(() {
          memoryUsageLabel = '''
External: ${profile.memoryUsage.externalUsage}
Heap: ${profile.memoryUsage.heapUsage}
''';
        });
      });
    };
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          children: <Widget>[
            RaisedButton(
              child: Text('CHECK ALLOC PROFILE (No GC)'),
              onPressed: sendLowMemory(false),
            ),
            RaisedButton(
              child: Text('CHECK ALLOC PROFILE (Try GC)'),
              onPressed: sendLowMemory(true),
            ),
            Text(memoryUsageLabel),
            RaisedButton(
              child: Text(imageButtonText),
              onPressed: loadingImage ? null : toggleImage,
            ),
            if (image != null)
              RawImage(
                image: image,
                width: 200,
                height: 200,
              )
            else
              Text('No image.'),
          ],
        ),
      ),
    );
  }
}
```

**Comment by dnfield** [comment]
You can also observe the expected memory behavior on the macOS embedding, which uses OpenGL. But I have verified that the same iOS embedding with OpenGL works as expected.

**Comment by dnfield** [comment]
Oh, and for whatever reason, this seems to be less of an issue in debug mode.

**Comment by kf6gpe** [comment]
@dnfield When you're back in the office, would you like to tackle this or delegate it? Asking for the weekly performance triage scan.

**Comment by dnfield** [comment]
I believe this is a skia issue currerently.  I can work on the Flutter end of it.

**Comment by chinmaygarde** [comment]
The updates in Skia seem to have been merged.

**Comment by dnfield** [comment]
Unfortunately, this seems to only have made the bad behvaior consistent between dev and profile. But I have not bisected to see if that was introduced by Dart or Skia.

**Comment by dnfield** [comment]
This has become a high priority item for customer: fun. See b/163533750

**Comment by chinmaygarde** [comment]
Per the internal update, this should be fixed when https://skia-review.googlesource.com/c/skia/+/310339 and https://skia-review.googlesource.com/c/skia/+/310762 are rolled in.

**Comment by dnfield** [comment]
Fixed in framework in https://github.com/flutter/flutter/pull/63979

**Comment by dnfield** [comment]
I'll leave this open to mirror internal bug until customer has verified.

**Comment by dnfield** [comment]
Roll was reverted for unrelated reasons. Should come in again on next engine roll.

**Comment by dnfield** [comment]
Fix relanded in 90aad51d3bc26aabad19897dd4a9b342fe823f7e

**Comment by dnfield** [comment]
I've marked the internal bug as fixed awaiting verification. Closing this, will re-open if internal verifies disagree.
